src/vis_cpu/vis_cpu.py

```diff
@@ -95,57 +115,88 @@ def vis_cpu(
-
+        for p1 in range(nax):
+            spl_axes = []
+            for p2 in range(nfeed):
+                spl_feeds = []
+                
+                # Loop over antennas
+                for i in range(nant):
+                    # Linear interpolation of primary beam pattern.
+                    spl = RectBivariateSpline(bm_pix_y, bm_pix_x, 
+                                              bm_cube[p1,p2,i], 
+                                              kx=1, ky=1)
+                    spl_feeds.append(spl)
+                spl_axes.append(spl_feeds)
+            splines.append(spl_axes)
+            
     # Loop over time samples
     for t, eq2top in enumerate(eq2tops.astype(real_dtype)):
         tx, ty, tz = crd_top = np.dot(eq2top, crd_eq)
-
+        
         # Primary beam response
         if beam_list is None:
             # Primary beam pattern using pixelized primary beam
             for i in range(nant):
-                A_s[i] = splines[i](ty, tx, grid=False)
-                # TODO: Try using a log-space beam for accuracy!
+                # Extract requested polarizations
+                for p1 in range(nax):
+                    for p2 in range(nfeed):
+                        A_s[p1,p2,i] = splines[p1][p2][i](ty, tx, grid=False)
         else:
             # Primary beam pattern using direct interpolation of UVBeam object
-            az, za = conversions.lm_to_az_za(tx, ty)
+            az, za = conversions.lm_to_az_za(tx, ty)       
             for i in range(nant):
                 interp_beam = beam_list[i].interp(az, za, np.atleast_1d(freq))[0]
-                A_s[i] = interp_beam[0, 0, 1]  # FIXME: assumes xx pol for now
-
+                
+                if polarized:
+                    A_s[:,:,i] = interp_beam[:,0,:,0,:] # spw=0 and freq=0
+                else:
+                    A_s[:,:,i] = interp_beam[0,0,1,:,:] # (phi, e) == 'xx' component
+        
+        # Horizon cut
         A_s = np.where(tz > 0, A_s, 0)
 
         # Calculate delays, where tau = (b * s) / c
         np.dot(antpos, crd_top, out=tau)
         tau /= c.value
-
-        # Component of complex phase factor for one antenna
-        # (actually, b = (antpos1 - antpos2) * crd_top / c; need dot product
+        
+        # Component of complex phase factor for one antenna 
+        # (actually, b = (antpos1 - antpos2) * crd_top / c; need dot product 
         # below to build full phase factor for a given baseline)
-        np.exp(1.0j * (ang_freq * tau), out=v)
-
+        np.exp(1.j * (ang_freq * tau), out=v)
+        
         # Complex voltages.
-        v *= A_s * Isqrt
+        v *= Isqrt
 
         # Compute visibilities using product of complex voltages (upper triangle).
+        # Input arrays have shape (Nax, Nfeed, [Nants], Npix
         for i in range(len(antpos)):
-            np.dot(v[i : i + 1].conj(), v[i:].T, out=vis[t, i : i + 1, i:])
-
-    return vis
+            vis[:, :, t, i:i+1, i:] = np.einsum(
```

Comment:
I think this is probably the only contentious line -- is the einsum the fastest way to do the dot product here, and can it be used on GPU?

Reply:
This requires some study. With this current einsum form, I'm hoping that there is a memory saving as a result of not having to expand the individual operations or keep intermediate steps. Avoiding einsum will become even more painful if we also want to allow Stokes Q,U sky models...

If this does turn out to be a bottleneck, a custom Cython or other accelerated function is probably the way to go.